crypto/kdf/tls1_prf.c

```diff
@@ -31,145 +32,127 @@ typedef struct {
     /* Buffer of concatenated seed data */
     unsigned char seed[TLS1_PRF_MAXBUF];
     size_t seedlen;
-} TLS1_PRF_PKEY_CTX;
+} TLS1_PRF_KDF_CTX;
 
-static int pkey_tls1_prf_init(EVP_PKEY_CTX *ctx)
+static void *kdf_tls1_prf_init(void)
 {
-    TLS1_PRF_PKEY_CTX *kctx;
-
-    if ((kctx = OPENSSL_zalloc(sizeof(*kctx))) == NULL) {
-        KDFerr(KDF_F_PKEY_TLS1_PRF_INIT, ERR_R_MALLOC_FAILURE);
-        return 0;
-    }
-    ctx->data = kctx;
+    TLS1_PRF_KDF_CTX *kctx;
 
-    return 1;
+    if ((kctx = OPENSSL_zalloc(sizeof(*kctx))) == NULL)
+        KDFerr(KDF_F_KDF_TLS1_PRF_INIT, ERR_R_MALLOC_FAILURE);
+    return kctx;
 }
 
-static void pkey_tls1_prf_cleanup(EVP_PKEY_CTX *ctx)
+static void kdf_tls1_prf_cleanup(void *ctx)
 {
-    TLS1_PRF_PKEY_CTX *kctx = ctx->data;
+    TLS1_PRF_KDF_CTX *kctx = ctx;
+
     OPENSSL_clear_free(kctx->sec, kctx->seclen);
     OPENSSL_cleanse(kctx->seed, kctx->seedlen);
     OPENSSL_free(kctx);
 }
 
-static int pkey_tls1_prf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+static int kdf_tls1_prf_ctrl(void *ctx, int type, int p1, void *p2)
 {
-    TLS1_PRF_PKEY_CTX *kctx = ctx->data;
+    TLS1_PRF_KDF_CTX *kctx = ctx;
+
     switch (type) {
-    case EVP_PKEY_CTRL_TLS_MD:
+    case EVP_KDF_CTRL_MD:
+        if (p2 == NULL)
+            return 0;
+
         kctx->md = p2;
         return 1;
 
-    case EVP_PKEY_CTRL_TLS_SECRET:
+    case EVP_KDF_CTRL_TLS_SECRET:
         if (p1 < 0)
             return 0;
-        if (kctx->sec != NULL)
-            OPENSSL_clear_free(kctx->sec, kctx->seclen);
+
+        OPENSSL_clear_free(kctx->sec, kctx->seclen);
         OPENSSL_cleanse(kctx->seed, kctx->seedlen);
         kctx->seedlen = 0;
         kctx->sec = OPENSSL_memdup(p2, p1);
         if (kctx->sec == NULL)
             return 0;
+
         kctx->seclen  = p1;
         return 1;
 
-    case EVP_PKEY_CTRL_TLS_SEED:
+    case EVP_KDF_CTRL_TLS_SEED:
```

Comment:
(excuse my lack of background knowledge here:) Is a TLS seed sufficiently different from a "salt"? (likewise is the secret sufficiently different from a "key"?): could they share the same define?

Reply:
I haven't checked to see if seed/salt can be considered equivalent here. Sometimes if the standard that defines an algorithm uses a particular terminology, it is good to be able to set a parameter using the same name as used in the standard - that way you know you are setting the parameter you think you are setting.
Also, sometimes algorithms that purport to solve the same problem are not interchangeable as they and their parameters have slightly different security properties; in that case it would be good for the parameters to have different names, even if at first glance they seem to be doing the same thing.

Reply:
From a binding perspective, it's good to have only a limited number of functions: if a new algorithm came out in a new openssl release that didn't require a different function call, then my users would instantly gain support.
However if every new algorithm requires some new `_set_egg` that is almost the same as `set_seed` then users have to wait for my bindings to update.

Reply:
New KDF algorithms do not add new function entry points to libcrypto.  `EVP_KDF_set1_salt()` and `EVP_KDF_add1_tls1_prf_seed()` are just macros provided for convenience that invoke the same function, `EVP_KDF_ctrl()`; you don't have to use them.
You could call either `EVP_KDF_ctrl()` or `EVP_KDF_ctrl_str()` directly if you want to avoid new bindings when new algorithms are added. In particular, `EVP_KDF_ctrl_str()` is designed to allow applications to supply parameters as strings (obtained from a config file or the command line, for example) so that new algorithms can be supported without even recompiling the application.

Note that in this case `EVP_KDF_set1_salt()` and `EVP_KDF_add1_tls1_prf_seed()` are different in at least one major way: `EVP_KDF_add1_tls1_prf_seed()` appends the bytes to the algorithm's internal buffer while `EVP_KDF_set1_salt()` completely replaces the existing value.

Reply:
All the parameter-setting macros are gone now, so from a binding perspective there are a limited number of functions and new ones won't be added as new algorithms are added.